integration-tests/src/test_loop/tests/fix_min_stake_ratio.rs

```diff
@@ -0,0 +1,138 @@
+        (0..8).map(|i| format!("account{}", i).parse().unwrap()).collect::<Vec<AccountId>>();
+    let clients = accounts.iter().cloned().collect_vec();
+    let small_validator = accounts[2].clone();
+
+    // Total stake is 62_500 NEAR - 1 unit, which is just enough for everyone
+    // to join the validator set after protocol upgrade.
+    let validators = vec![
+        near_primitives::types::AccountInfo {
+            account_id: accounts[0].clone(),
+            public_key: near_primitives::test_utils::create_test_signer(accounts[0].as_str())
+                .public_key(),
+            amount: 31_249 * ONE_NEAR,
+        },
+        near_primitives::types::AccountInfo {
+            account_id: accounts[1].clone(),
+            public_key: near_primitives::test_utils::create_test_signer(accounts[1].as_str())
+                .public_key(),
+            amount: 31_250 * ONE_NEAR - 1,
+        },
+        near_primitives::types::AccountInfo {
+            account_id: accounts[2].clone(),
+            public_key: near_primitives::test_utils::create_test_signer(accounts[2].as_str())
+                .public_key(),
+            amount: ONE_NEAR,
+        },
+    ];
+
+    // Create chain with version before FixMinStakeRatio was enabled.
+    // Check that the small validator is not included in the validator set.
+    let mut genesis_builder = TestGenesisBuilder::new();
+    genesis_builder
+        .genesis_time_from_clock(&builder.clock())
+        .shard_layout(ShardLayout::get_simple_nightshade_layout_v3())
+        .protocol_version(ProtocolFeature::FixMinStakeRatio.protocol_version() - 1)
+        .epoch_length(epoch_length)
+        .validators_raw(validators, 1, 2)
+        // For genesis, set high minimum stake ratio so that small validator
+        // will be excluded from the validator set.
+        .minimum_stake_ratio(Rational32::new(1, 6_250))
+        // Disable validator rewards.
+        .max_inflation_rate(Rational32::new(0, 1));
+    for account in &accounts {
+        genesis_builder.add_user_account_simple(account.clone(), initial_balance);
+    }
+    let genesis = genesis_builder.build();
+
+    let TestLoopEnv { mut test_loop, datas: node_datas, tempdir } =
+        builder.genesis(genesis).clients(clients).build();
+
+    let client_sender = node_datas[0].client_sender.clone();
+    let client_handle = node_datas[0].client_sender.actor_handle();
+    let initial_validators = get_epoch_all_validators(&test_loop.data.get(&client_handle).client);
+    assert_eq!(initial_validators.len(), 2);
+    assert!(!initial_validators.contains(&small_validator.to_string()));
+
+    // Run chain for couple epochs.
+    // Due to how protocol version voting works, chain must automatically
+    // upgrade to the latest protocol version which includes FixMinStakeRatio
+    // enabled.
+    // In the epoch where it happens, small validator must join the validator
+    // set.
+    let latest_epoch_height = AtomicU64::new(0);
```

Comment:
is atomic needed here? I thought testloop is single threaded?

Reply:
It is needed because TestLoop doesn't currently support "condition checkers" which mutate state. I didn't find other way to get around that.
Instead, separated this into function `stake_if_new_epoch_started`, PTAL.